### PR TITLE
#121962 - Add ListenableFutureBuilder

### LINKED ISSUE #121962: Introduce ListenableFutureBuilder to Simplify Using Controllers in the Composition of a StatelessWidget
## Use case

I should be able to instantiate a `Listenable` controller with async work and use this controller to trigger rebuilds without using some external means to hold onto the `Listenable` variable. I also shouldn't have to subtype `StatefulWidget` to do this. I should be able to use straightforward composition to achieve this. I've created a widget that solves this problem, and there is a draft PR for it. But, before I jump to that, I will try to explain the existing problem.

## Existing Problem
On its face, this seems like something that would already be straightforward in Flutter, but it isn't. If you have a controller that extends `ValueNotifier` or `ChangeNotifier`, there is no straightforward way to compose a widget that listens to this without subtyping `StatefulWidget`, using a top-level singleton or something like Provider. Let's take a look at a simple example with `AnimatedBuilder` and `ValueNotifier`.

This example is obviously wrong because the parent widget could reinstantiate `MyHomePage`.

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) => MaterialApp(
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: MyHomePage(),
      );
}

class MyHomePage extends StatelessWidget {
  MyHomePage({super.key});

  //A StatelessWidget should never hold onto state in this way
  ValueNotifier<int> _counter = ValueNotifier<int>(0);

  @override
  Widget build(BuildContext context) => AnimatedBuilder(
        animation: _counter,
        builder: (context, child) => Scaffold(
          appBar: AppBar(),
          body: Center(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                const Text(
                  'You have pushed the button this many times:',
                ),
                Text(
                  '${_counter.value}',
                  style: Theme.of(context).textTheme.headlineMedium,
                ),
              ],
            ),
          ),
          floatingActionButton: FloatingActionButton(
            onPressed: () => _counter.value++,
            tooltip: 'Increment',
            child: const Icon(Icons.add),
          ),
        ),
      );
}
```

The above works well but it violates a fundamental principle of Flutter: `StatelessWidget`s should not hold onto state. This is because the parent widget could be reinstantiated, and the child widget would lose its state. This is why `StatefulWidget` exists. It allows the child widget to hold onto its state. So, we have a couple of choices:

- Turn this into a `StatefulWidget` and move the controller to the `State`
- Declare the controller as a top-level singleton
- Use something like Provider to hold onto the variable and pass it down to the child widget

None of these options are elegant solutions to the problem, and the problem gets worse when we consider async work. They all add complexity to the codebase and make it harder to reason about. 

Let's expand on the existing, already problematic example with async work.

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) => MaterialApp(
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: MyHomePage(),
      );
}

class MyHomePage extends StatelessWidget {
  MyHomePage({super.key});

  @override
  Widget build(BuildContext context) => FutureBuilder<ValueNotifier<int>>(
        //This is a problem because even though the FutureBuilder will store the result in 
        //State, if the parent rebuilds, this function will execute many times
        future: getController(),
        builder: (context, snapshot) => AnimatedBuilder(
          animation: snapshot.data ?? ValueNotifier<int>(0),
          builder: (context, child) => Scaffold(
            appBar: AppBar(),
            body: Center(
                child: snapshot.connectionState == ConnectionState.done
                    ? Column(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: <Widget>[
                          const Text(
                            'You have pushed the button this many times:',
                          ),
                          Text(
                            '${snapshot.data!.value}',
                            style: Theme.of(context).textTheme.headlineMedium,
                          ),
                        ],
                      )
                    : const CircularProgressIndicator.adaptive()),
            floatingActionButton: FloatingActionButton(
              onPressed: () => snapshot.data?.value++,
              tooltip: 'Increment',
              child: const Icon(Icons.add),
            ),
          ),
        ),
      );
}

Future<ValueNotifier<int>> getController() async =>
    Future.delayed(const Duration(seconds: 2), () => ValueNotifier<int>(0));
```

Again, this example works, but it has a problem. If the parent widget reinstantiates this widget, the async function will execute again, causing the controller to be reinstantiated. Again, the controller resets. You can also see that this version is already getting more complicated and harder to read.

Take a look at the `future` property of the `FutureBuilder`. This is problematic because it's not a function. It forces the composition to reference a declaration or make a function call to get the `Future`.  Every time the parent widget rebuilds, the `FutureBuilder` will reference the declaration or call the function again. This recently caused a situation where the Flutter team had to redo a YouTube video for clarification. In an ideal world, the FutureBuilder would hold onto the value from the `Future`, as it does, but only call a function in a lazy way when it needs that value. 

## Proposal

The good news is that there is an easy way to solve several of these issues in one hit and massively simplify using controllers in the composition of a  `StatelessWidget`. 

Introducing `ListenableFutureBuilder`:

```dart
import 'package:flutter/material.dart';
import 'package:flutter_application_4/listenable_future_builder.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) => MaterialApp(
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: const MyHomePage(),
      );
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({super.key});

  @override
  Widget build(BuildContext context) =>
      ListenableFutureBuilder<ValueNotifier<int>>(
        //This function is safe because it will only execute once
        //ListenableFutureBuilder makes the first call, it will not execute the call again
        listenable: getController,
        builder: (context, child, snapshot) => Scaffold(
          appBar: AppBar(),
          body: Center(
              child: snapshot.connectionState == ConnectionState.done
                  ? Column(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: <Widget>[
                        const Text(
                          'You have pushed the button this many times:',
                        ),
                        Text(
                          '${snapshot.data!.value}',
                          style: Theme.of(context).textTheme.headlineMedium,
                        ),
                      ],
                    )
                  : const CircularProgressIndicator.adaptive()),
          floatingActionButton: FloatingActionButton(
            onPressed: () => snapshot.data?.value++,
            tooltip: 'Increment',
            child: const Icon(Icons.add),
          ),
        ),
      );
}

Future<ValueNotifier<int>> getController() async =>
    Future.delayed(const Duration(seconds: 2), () => ValueNotifier<int>(0));
```

This widget simplifies the code from the previous two examples, but more importantly, it solves the problem of the controller being reinstantiated when the parent widget is reinstantiated. It does this by calling `listenable` function only once and then listening to the `Listenable` returned by that function. `ListenableFutureBuilder` is a `StatefulWidget` and it holds onto the `Listenable` controller. This means that the controller will not be reinstantiated when the parent widget is reinstantiated. The result is a large simplification, but also removes dependence on things like Provider to ferry the controller in via an inherited widget.

### PR DESCRIPTION
**Pull request by MelbourneDeveloper**
This PR adds the `ListenableFutureBuilder` widget and tests. Briefly: this widget allows you to compose `StatelessWidget`s with async work, and without the need to extend StatefulWidget, or use external dependency managers. For a full description of the issue and solution see #121962

This PR fixes #121962

See a live sample [here](https://dartpad.dev/?id=3da064ae821e85b20c74c78b255aa03f)

**This is a draft PR and  I will revisit this, and address issues as they come up. I will go over the criteria here again before full submission**

## Pre-launch Checklist

- [✅] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [✅ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [✅ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [✅ ] I signed the [CLA].
- [✅ ] I listed at least one issue that this PR fixes in the description above.
- [✅] I updated/added relevant documentation (doc comments with `///`).
- [✅ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [✅ ] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/listenable_future_builder.dart [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    _unsubscribe();
+    _snapshot = AsyncSnapshot<T>.nothing();
+    super.dispose();
+  }
+
+  void _handleChange() => setState(() {});
+
+  void _subscribe() {
+    final Object callbackIdentity = Object();
+    _activeCallbackIdentity = callbackIdentity;
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/listenable_future_builder.dart:198]: What scenario is this guarding against?

File: packages/flutter/test/widgets/listenable_future_builder_test.dart [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        home: ListenableFutureBuilder<ValueNotifier<int>>(
+          listenable: () async => ValueNotifier<int>(42),
+          builder: (BuildContext context, Widget? child,
+                  AsyncSnapshot<ValueNotifier<int>> snapshot) =>
+              snapshot.connectionState == ConnectionState.done
+                  ? Text('${snapshot.data!.value}')
+                  : const CircularProgressIndicator(),
+        ),
+      ),
+    );
```
Comment (goderbauer) [inline packages/flutter/test/widgets/listenable_future_builder_test.dart:22]: nit: formatting

Comment (goderbauer) [reply packages/flutter/test/widgets/listenable_future_builder_test.dart:22]: (here and elsewhere in this file)

File: packages/flutter/test/widgets/listenable_future_builder_test.dart [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+    // Increment the value of the notifier
+    notifier.value = 1;
+
+    // Verify that Text widget with value 1 is shown after notifier is changed
+    await tester.pumpAndSettle();
+    expect(find.text('1'), findsOneWidget);
+  });
+
+  ///Notes: we need to confirm that the builder drops its reference to the
```
[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/test/widgets/listenable_future_builder_test.dart:171]: Also, these should just be `//`, not `///`

**Comment by goderbauer** [comment]
Looks like the analyzer is also unhappy with this change. Could you take a look?

**Comment by MelbourneDeveloper** [comment]
Hi @goderbauer 

I noticed that there were some issues there but was waiting for other feedback before I investigate. Is there any other feedback before I try to figure out what's going wrong here? 

There is a lot of detail on the issue itself. 

Thanks

**Comment by goderbauer** [comment]
I left a comment on the issue as well about my high-level concerns with this setup: https://github.com/flutter/flutter/issues/121962#issuecomment-1485726048

**Comment by MelbourneDeveloper** [comment]
Hi @goderbauer. Thanks for the detailed responses here. Sorry, I missed the activity here. I didn't realize you had already reviewed this in detail. I'm closing this draft PR because this was more of a demonstration of how the issue could be solved rather than a finished PR. 

Regarding the issue itself, it doesn't appear that my suggestion is clear. I would hope to leave that open and allow for further discussion.

**Pull request closed without merging**